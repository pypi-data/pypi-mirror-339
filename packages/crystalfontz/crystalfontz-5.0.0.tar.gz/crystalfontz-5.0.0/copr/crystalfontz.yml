apiVersion: coprctl/v1alpha1
kind: package-scm
metadata:
  name: crystalfontz
spec:
  auto_rebuild: true
  ownername: jfhbrook
  projectname: joshiverse
  clone-url: https://github.com/jfhbrook/crystalfontz
  commit: null
  spec: crystalfontz.spec
  type: git
  method: tito
