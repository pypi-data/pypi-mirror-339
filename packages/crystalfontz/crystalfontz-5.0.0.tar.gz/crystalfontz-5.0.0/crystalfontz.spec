Name: crystalfontz
Version: 5.0.0
Release: 1
License: MPL-2.0
Summary: Serial client and Linux service for Crystalfontz LCD displays

URL: https://github.com/jfhbrook/crystalfontz
Source0: %{name}-%{version}.tar.gz
BuildArch: noarch

Requires: python-crystalfontz
Requires: python-sdbus

%description


%prep
%autosetup


%build
tar -xzf %{SOURCE0}


%install
mkdir -p %{buildroot}%{_prefix}/lib/systemd/system
mkdir -p %{buildroot}%{_bindir}
install -p -D -m 0644 systemd/crystalfontz.service %{buildroot}%{_prefix}/lib/systemd/system/crystalfontz.service
install -p -D -m 0644 dbus/org.jfhbrook.crystalfontz.conf %{buildroot}%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
install -p -m 755 bin/crystalfontz-dbus %{buildroot}%{_bindir}/crystalfontz

%check


%files
%{_prefix}/lib/systemd/system/crystalfontz.service
%{_prefix}/share/dbus-1/system.d/org.jfhbrook.crystalfontz.conf
%{_bindir}/crystalfontz

%changelog
* Tue Apr 08 2025 Josh Holbrook <josh.holbrook@gmail.com> 5.0.0-1
- API Changes:
  - **NEW:** `Receiver` class
    - A subclass of `asyncio.Queue`
  - `Client`:
    - **BREAKING:** `subscribe` and `unsubscribe` methods use new `Receiver` class
    - Emit unmatched exceptions on expecting receivers instead of resolving `client.closed`
    - `detect_baud_rate` exposes `timeout` and `retry_times` arguments
    - Document `detect_baud_rate`
  - **BREAKING:** `ClientProtocol`/`EffectClient`
    - `crystalfontz.protocol.ClientProtocol` type has been replaced by `crystalfontz.effects.EffectClient`
    - `EffectClient` enforces a smaller API than `ClientProtocol` did previously
  - **BREAKING:** `Response`:
    - `Response.from_bytes` accepts bytes as from packets, rather than `__init__`
    - `Response.__init__` accepts properties as arguments
  - **BREAKING:** `SpecialCharacter` API:
    - Rename `as_bytes` method to `to_bytes`
    - Store pixels as `List[List[bool]]` instead of `List[List[int]]`
  - `KeyState` includes `keypress: KeyPress` attribute
- CLI changes:
  - **BREAKING:** CLI now invoked through `python3 -m crystalfontz`
    - Optional unpackaged `crystalfontz` entry point at `./bin/crystalfontz`
  - Use `configurence` library
  - Respect `CRYSTALFONTZ_CONFIG_FILE` environment variable
  - Improve error reporting for timeouts
  - Client now respects CLI settings
    - Cause was the `Client` constructor being called twice
  - Additional commands accept bytes as arguments
    - `python3 -m crystalfontz line 1`
    - `python3 -m crystalfontz line 2`
    - `python3 -m crystalfontz send`
  - Help for `python3 -m crystalfontz listen` `--for` option
- **NEW:** Dbus support:
  - `crystalfontz.dbus.DbusInterface` dbus Interface class, implementing most commands
  - `crystalfontz.dbus.DbusClient` dbus client class
  - `crystalfontz.dbus.domain` API for mapping domain objects to dbus types
  - `python3 -m crystalfontz.dbus.service` dbus service CLI
    - Optional unpackaged `crystalfontz.dbus.service` entry point at `./bin/crystalfontz-service`
  - `python3 -m crystalfontz.dbus.client` dbus client CLI
    - Optional unpackaged `crystalfontz.dbus.client` entry point at `./bin/crystalfontz-dbus`
- Packaging and Releases:
  - `python-crystalfontz` COPR package spec
    - This has been moved from <https://github.com/jfhbrook/public>
  - **NEW:** `crystalfontz` COPR package
    - Depends on `python-crystalfontz` COPR package
    - Includes systemd unit for `python3 -m crystalfontz.dbus.service`
    - **BREAKING:** Includes shim bin `crystalfontz` -> `python3 -m crystalfontz.dbus.client`
  - Tito based release tagging
  - GitHub release
  - Improved PyPI classifiers
  - **BREAKING:** Release under MPL-2.0 license
- Integration tests:
  - Use `python-gak` plugin
  - Use snapshots
  - Leverage a config file at `./tests/fixtures/crystalfontz.yaml`
    - Can be overridden with `CRYSTALFONTZ_CONFIG_FILE` environment variable


