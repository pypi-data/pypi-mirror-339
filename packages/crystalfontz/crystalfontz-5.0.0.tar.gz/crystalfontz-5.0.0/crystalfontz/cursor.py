from enum import Enum


class CursorStyle(Enum):
    """
    A cursor style, as set with command 12 (0x0C): Set LCD Cursor Style.

    - **NONE**:  No cursor.
    - **BLINKING_BLOCK**: Blinking block cursor.
    - **STATIC_UNDERSCORE**: Static underscore cursor.
    - **BLINKING_UNDERSCORE**: Blinking underscore cursor. On the CFA633, this
      represents a blinking block plus an underscore.
    """

    NONE = 0
    BLINKING_BLOCK = 1
    STATIC_UNDERSCORE = 2
    BLINKING_UNDERSCORE = 3
