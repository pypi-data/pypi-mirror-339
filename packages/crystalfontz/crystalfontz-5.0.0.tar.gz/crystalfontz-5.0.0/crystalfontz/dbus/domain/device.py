from typing import ClassVar, Tuple

from crystalfontz.dbus.domain.base import (
    BytesM,
    ModelM,
    ModelT,
    RevisionM,
    RevisionT,
    struct,
    t,
    Uint16M,
    Uint16T,
)
from crystalfontz.device import Device, DeviceStatus, lookup_device

WidthT = Uint16T


class WidthM(Uint16M):
    pass


HeightT = Uint16T


class HeightM(Uint16M):
    pass


DeviceT = Tuple[ModelT, RevisionT, RevisionT, WidthT, HeightT, WidthT, HeightT, Uint16T]


class DeviceM:
    t: ClassVar[str] = struct(
        ModelM, RevisionM, RevisionM, WidthM, HeightM, WidthM, HeightM, Uint16M
    )

    @staticmethod
    def pack(device: Device) -> DeviceT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            device.lines,
            device.columns,
            device.character_width,
            device.character_height,
            device.n_temperature_sensors,
        )

    @staticmethod
    def unpack(device: DeviceT) -> Device:
        model = device[0]
        hardware_rev = device[1]
        firmware_rev = device[2]

        return lookup_device(model, hardware_rev, firmware_rev)


DeviceStatusT = Tuple[ModelT, RevisionT, RevisionT, bytes]


class DeviceStatusM:
    t: ClassVar[str] = t(ModelM, RevisionM, RevisionM, BytesM)

    @staticmethod
    def pack(status: DeviceStatus, device: Device) -> DeviceStatusT:
        return (
            device.model,
            device.hardware_rev,
            device.firmware_rev,
            status.to_bytes(device),
        )

    @staticmethod
    def unpack(status: DeviceStatusT) -> DeviceStatus:
        model, hardware_rev, firmware_rev, data = status
        device = lookup_device(model, hardware_rev, firmware_rev)
        return device.status(data)
