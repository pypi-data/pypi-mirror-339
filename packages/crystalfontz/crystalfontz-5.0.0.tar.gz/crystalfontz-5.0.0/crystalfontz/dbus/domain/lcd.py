from typing import ClassVar, Dict

from crystalfontz.lcd import LcdRegister

LcdRegisterT = bool

LCD_REGISTERS: Dict[bool, LcdRegister] = {
    bool(register.value): register for register in LcdRegister
}


class LcdRegisterM:
    """
    Map `LcdRegister` to and from `LcdRegisterT` (`bool`).
    """

    t: ClassVar[str] = "b"

    @staticmethod
    def pack(register: LcdRegister) -> LcdRegisterT:
        return bool(register.value)

    @staticmethod
    def unpack(register: LcdRegisterT) -> LcdRegister:
        return LCD_REGISTERS[register]
