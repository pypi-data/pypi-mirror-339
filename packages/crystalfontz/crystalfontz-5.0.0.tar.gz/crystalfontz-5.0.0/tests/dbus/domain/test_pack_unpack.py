from typing import Any, Callable, cast, List, Optional

import pytest

from tests.helpers import special_character_as_rows

from crystalfontz.atx import AtxPowerSwitchFunction
from crystalfontz.baud import FAST_BAUD_RATE, SLOW_BAUD_RATE
from crystalfontz.character import SMILEY_FACE
from crystalfontz.config import Config
from crystalfontz.cursor import CursorStyle
from crystalfontz.dbus.domain.atx import (
    AtxPowerSwitchFunctionalitySettingsM,
    AtxPowerSwitchFunctionM,
)
from crystalfontz.dbus.domain.base import (
    OptBytesM,
    OptFloatM,
    OptIntM,
    RetryTimesM,
    TimeoutM,
)
from crystalfontz.dbus.domain.baud import BaudRateM
from crystalfontz.dbus.domain.command import (
    ConfigureKeyReportingM,
    ConfigureWatchdogM,
    DowTransactionM,
    PingM,
    ReadDowDeviceInformationM,
    ReadGpioM,
    ReadLcdMemoryM,
    SendCommandToLcdControllerM,
    SendDataM,
    SetAtxPowerSwitchFunctionalityM,
    SetBacklightM,
    SetBaudRateM,
    SetContrastM,
    SetCursorPositionM,
    SetCursorStyleM,
    SetGpioM,
    SetLineM,
    SetupLiveTemperatureDisplayM,
    SetupTemperatureReportingM,
    SimpleCommandM,
    SpecialCharacterM,
    WriteUserFlashAreaM,
)
from crystalfontz.dbus.domain.config import ConfigM
from crystalfontz.dbus.domain.cursor import CursorStyleM
from crystalfontz.dbus.domain.device import DeviceM
from crystalfontz.dbus.domain.gpio import GpioSettingsM, OptGpioSettingsM
from crystalfontz.dbus.domain.keys import (
    KeypadBrightnessM,
    KeyPressM,
)
from crystalfontz.dbus.domain.lcd import LcdRegisterM
from crystalfontz.dbus.domain.response import (
    DowDeviceInformationM,
    DowTransactionResultM,
    GpioReadM,
    KeyActivityReportM,
    KeypadPolledM,
    LcdMemoryM,
    PongM,
    TemperatureReportM,
    VersionsM,
)
from crystalfontz.dbus.domain.temperature import (
    TemperatureDigitsM,
    TemperatureDisplayItemM,
)
from crystalfontz.device import CFA533
from crystalfontz.gpio import GpioDriveMode, GpioFunction, GpioSettings, GpioState
from crystalfontz.keys import (
    KeyActivity,
    KeyState,
    KeyStates,
    KP_DOWN,
    KP_ENTER,
    KP_EXIT,
    KP_LEFT,
    KP_RIGHT,
    KP_UP,
)
from crystalfontz.lcd import LcdRegister
from crystalfontz.response import (
    DowDeviceInformation,
    DowTransactionResult,
    GpioRead,
    KeyActivityReport,
    KeypadPolled,
    LcdMemory,
    Pong,
    TemperatureReport,
    Versions,
)
from crystalfontz.temperature import TemperatureDisplayItem, TemperatureUnit

ValidateFn = Callable[[Any, Any], None]


def validate_is(actual: Any, expected: Any) -> None:
    assert isinstance(actual, expected.__class__)


def validate_gpio_settings(actual: Any, expected: Any) -> None:
    if expected is None:
        assert actual is None
        return
    assert actual.function == expected.function
    assert actual.mode == expected.mode


def validate_key_activity_report(actual: Any, expected: Any) -> None:
    assert actual.activity == expected.activity


def validate_temperature_report(actual: Any, expected: Any) -> None:
    assert actual.index == expected.index
    assert abs(actual.celsius - expected.celsius) < 0.001, "celsius matches"
    assert abs(actual.fahrenheit - expected.fahrenheit) < 0.001, "fahrenheit matches"


@pytest.mark.parametrize(
    "entity,map_cls,validate",
    [
        #
        # Base types
        #
        (1, OptIntM, None),
        (None, OptIntM, None),
        (1.0, OptFloatM, None),
        (None, OptFloatM, None),
        (b"hello", OptBytesM, None),
        (None, OptBytesM, None),
        (1.0, TimeoutM, None),
        (None, TimeoutM, None),
        (1, RetryTimesM, None),
        (None, RetryTimesM, None),
        #
        # Complex entities
        #
        (
            cast(Any, Config)(
                file="/etc/crystalfontz.yaml",
                port="/dev/ttyUSB1",
                model="CFA533",
                hardware_rev="h1.4",
                firmware_rev="u1v2",
                baud_rate=FAST_BAUD_RATE,
                timeout=0.250,
                retry_times=1,
            ),
            ConfigM,
            None,
        ),
        (CursorStyle.BLINKING_UNDERSCORE, CursorStyleM, None),
        (CFA533(), DeviceM, validate_is),
        (
            GpioSettings(
                function=GpioFunction.UNUSED,
                up=GpioDriveMode.FAST_STRONG,
                down=GpioDriveMode.RESISTIVE,
            ),
            GpioSettingsM,
            validate_gpio_settings,
        ),
        (
            GpioSettings(
                function=GpioFunction.UNUSED,
                up=GpioDriveMode.FAST_STRONG,
                down=GpioDriveMode.RESISTIVE,
            ),
            OptGpioSettingsM,
            validate_gpio_settings,
        ),
        (
            None,
            OptGpioSettingsM,
            validate_gpio_settings,
        ),
        (1.0, KeypadBrightnessM, None),
        (None, KeypadBrightnessM, None),
        (LcdRegister.DATA, LcdRegisterM, None),
        (LcdRegister.CONTROL, LcdRegisterM, None),
        (
            TemperatureDisplayItem(
                index=1, n_digits=3, row=1, column=1, units=TemperatureUnit.CELSIUS
            ),
            TemperatureDisplayItemM,
            None,
        ),
        #
        # Responses
        #
        (DowDeviceInformation(0x00, 0xFF), DowDeviceInformationM, validate_is),
        (DowTransactionResult(0, b"\00", 0xFF), DowTransactionResultM, validate_is),
        (
            GpioRead(
                0,
                GpioState(state=True, falling=False, rising=False),
                5,
                GpioSettings(
                    function=GpioFunction.UNUSED,
                    up=GpioDriveMode.FAST_STRONG,
                    down=GpioDriveMode.RESISTIVE,
                ),
            ),
            GpioReadM,
            validate_is,
        ),
        (
            KeypadPolled(
                KeyStates(
                    up=KeyState(KP_UP, True, False, False),
                    enter=KeyState(KP_ENTER, True, False, False),
                    exit=KeyState(KP_EXIT, True, False, False),
                    left=KeyState(KP_LEFT, True, False, False),
                    right=KeyState(KP_RIGHT, True, False, False),
                    down=KeyState(KP_DOWN, True, False, False),
                )
            ),
            KeypadPolledM,
            validate_is,
        ),
        (LcdMemory(0x00, b"\00"), LcdMemoryM, validate_is),
        (Pong(b"ping"), PongM, validate_is),
        (Versions("CFA533", "h1.4", "u1v2"), VersionsM, validate_is),
        #
        # Reports
        #
        (
            KeyActivityReport(KeyActivity.KEY_UP_PRESS),
            KeyActivityReportM,
            validate_key_activity_report,
        ),
        (
            TemperatureReport(1, 30.0, 90.0),
            TemperatureReportM,
            validate_temperature_report,
        ),
    ],
)
def test_domain_pack_unpack(
    entity: Any, map_cls: Any, validate: Optional[ValidateFn], snapshot
) -> None:
    packed = map_cls.pack(entity)

    assert packed == snapshot

    if hasattr(map_cls, "unpack"):
        unpacked = map_cls.unpack(packed)
        if validate:
            validate(unpacked, entity)
        else:
            assert unpacked == entity


def test_special_character_pack_unpack(snapshot) -> None:
    device = CFA533()
    packed = SpecialCharacterM.pack(SMILEY_FACE, device)

    assert packed == snapshot

    unpacked = SpecialCharacterM.unpack(packed)

    assert special_character_as_rows(unpacked) == special_character_as_rows(SMILEY_FACE)


@pytest.mark.parametrize(
    "packed,map_cls",
    [
        #
        # Mostly entities that only get unpacked
        #
        (b"", OptBytesM),
        (
            ([AtxPowerSwitchFunction.KEYPAD_RESET.value], False, True, True, 1.0),
            AtxPowerSwitchFunctionalitySettingsM,
        ),
        (SLOW_BAUD_RATE, BaudRateM),
        (FAST_BAUD_RATE, BaudRateM),
        (KP_UP, KeyPressM),
        (3, TemperatureDigitsM),
        (5, TemperatureDigitsM),
    ],
)
def test_domain_unpack_pack(packed: Any, map_cls: Any, snapshot) -> None:
    entity = map_cls.unpack(packed)

    assert entity == snapshot

    if hasattr(map_cls, "pack"):
        repacked = map_cls.pack(entity)
        assert repacked == packed


@pytest.mark.parametrize(
    "packed,map_cls",
    [
        #
        # Commands. These never get packed, and they take multiple arguments
        #
        ([], SimpleCommandM),
        ([b"hello"], PingM),
        ([b"data"], WriteUserFlashAreaM),
        ([b"hello"], SetLineM),
        ([0x01], ReadLcdMemoryM),
        ([1, 1], SetCursorPositionM),
        ([CursorStyle.BLINKING_UNDERSCORE.value], SetCursorStyleM),
        ([0.5], SetContrastM),
        ([0.5, 0.5], SetBacklightM),
        ([0.5, -1.0], SetBacklightM),
        ([1], ReadDowDeviceInformationM),
        ([[1, 2, 3]], SetupTemperatureReportingM),
        ([1, 3, [0x00]], DowTransactionM),
        ([1, 3, []], DowTransactionM),
        ([1, [1, 3, 1, 1, True]], SetupLiveTemperatureDisplayM),
        ([False, 0x01], SendCommandToLcdControllerM),
        ([[KP_UP], [KP_UP]], ConfigureKeyReportingM),
        (
            [([AtxPowerSwitchFunction.KEYPAD_POWER_OFF.value], False, True, True, 1.5)],
            SetAtxPowerSwitchFunctionalityM,
        ),
        ([2], ConfigureWatchdogM),
        ([1, 5, [0x00]], SendDataM),
        ([FAST_BAUD_RATE], SetBaudRateM),
        (
            [
                1,
                10,
                OptGpioSettingsM.pack(
                    GpioSettings(
                        function=GpioFunction.UNUSED,
                        up=GpioDriveMode.FAST_STRONG,
                        down=GpioDriveMode.RESISTIVE,
                    ),
                ),
            ],
            SetGpioM,
        ),
        ([1], ReadGpioM),
    ],
)
def test_domain_unpack_command(packed: List[Any], map_cls: Any, snapshot) -> None:
    with_timeout_retry = map_cls.unpack(*(packed + [0.250, 3]))

    assert with_timeout_retry == snapshot

    without_timeout_retry = map_cls.unpack(
        *(packed + [TimeoutM.none, RetryTimesM.none])
    )

    assert without_timeout_retry == snapshot


@pytest.mark.parametrize(
    "packed,map_cls",
    [
        #
        # Entities which have validation logic on unpack
        #
        (0x00, AtxPowerSwitchFunctionM),
        (12, BaudRateM),
        (42, KeyPressM),
        (4, TemperatureDigitsM),
    ],
)
def test_domain_unpack_error(packed: Any, map_cls: Any) -> None:
    with pytest.raises(ValueError):
        map_cls.unpack(packed)
