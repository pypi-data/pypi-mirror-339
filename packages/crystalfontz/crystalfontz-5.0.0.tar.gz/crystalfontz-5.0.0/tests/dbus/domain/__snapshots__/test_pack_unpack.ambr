# serializer version: 1
# name: test_domain_pack_unpack[1-OptIntM-None]
  1
# ---
# name: test_domain_pack_unpack[1-RetryTimesM-None]
  1
# ---
# name: test_domain_pack_unpack[1.0-KeypadBrightnessM-None]
  1.0
# ---
# name: test_domain_pack_unpack[1.0-OptFloatM-None]
  1.0
# ---
# name: test_domain_pack_unpack[1.0-TimeoutM-None]
  1.0
# ---
# name: test_domain_pack_unpack[CursorStyle.BLINKING_UNDERSCORE-CursorStyleM-None]
  3
# ---
# name: test_domain_pack_unpack[LcdRegister.CONTROL-LcdRegisterM-None]
  True
# ---
# name: test_domain_pack_unpack[LcdRegister.DATA-LcdRegisterM-None]
  False
# ---
# name: test_domain_pack_unpack[None-KeypadBrightnessM-None]
  -1.0
# ---
# name: test_domain_pack_unpack[None-OptBytesM-None]
  b''
# ---
# name: test_domain_pack_unpack[None-OptFloatM-None]
  -1.0
# ---
# name: test_domain_pack_unpack[None-OptGpioSettingsM-validate_gpio_settings]
  65280
# ---
# name: test_domain_pack_unpack[None-OptIntM-None]
  -1
# ---
# name: test_domain_pack_unpack[None-RetryTimesM-None]
  -1
# ---
# name: test_domain_pack_unpack[None-TimeoutM-None]
  -1.0
# ---
# name: test_domain_pack_unpack[entity10-ConfigM-None]
  tuple(
    '/etc/crystalfontz.yaml',
    '/dev/ttyUSB1',
    'CFA533',
    'h1.4',
    'u1v2',
    115200,
    0.25,
    1,
  )
# ---
# name: test_domain_pack_unpack[entity12-DeviceM-validate_is]
  tuple(
    'CFA533',
    'h1.4',
    'u1v2',
    2,
    16,
    6,
    8,
    32,
  )
# ---
# name: test_domain_pack_unpack[entity13-GpioSettingsM-validate_gpio_settings]
  0
# ---
# name: test_domain_pack_unpack[entity14-OptGpioSettingsM-validate_gpio_settings]
  0
# ---
# name: test_domain_pack_unpack[entity20-TemperatureDisplayItemM-None]
  tuple(
    1,
    3,
    1,
    1,
    False,
  )
# ---
# name: test_domain_pack_unpack[entity21-DowDeviceInformationM-validate_is]
  tuple(
    0,
    255,
  )
# ---
# name: test_domain_pack_unpack[entity22-DowTransactionResultM-validate_is]
  tuple(
    0,
    b'\x00',
    255,
  )
# ---
# name: test_domain_pack_unpack[entity23-GpioReadM-validate_is]
  tuple(
    0,
    tuple(
      True,
      False,
      False,
    ),
    5,
    0,
  )
# ---
# name: test_domain_pack_unpack[entity24-KeypadPolledM-validate_is]
  tuple(
    tuple(
      True,
      False,
      False,
    ),
    tuple(
      True,
      False,
      False,
    ),
    tuple(
      True,
      False,
      False,
    ),
    tuple(
      True,
      False,
      False,
    ),
    tuple(
      True,
      False,
      False,
    ),
    tuple(
      True,
      False,
      False,
    ),
  )
# ---
# name: test_domain_pack_unpack[entity25-LcdMemoryM-validate_is]
  tuple(
    0,
    b'\x00',
  )
# ---
# name: test_domain_pack_unpack[entity26-PongM-validate_is]
  b'ping'
# ---
# name: test_domain_pack_unpack[entity27-VersionsM-validate_is]
  tuple(
    'CFA533',
    'h1.4',
    'u1v2',
  )
# ---
# name: test_domain_pack_unpack[entity28-KeyActivityReportM-validate_key_activity_report]
  1
# ---
# name: test_domain_pack_unpack[entity29-TemperatureReportM-validate_temperature_report]
  tuple(
    1,
    30.0,
    90.0,
  )
# ---
# name: test_domain_pack_unpack[hello-OptBytesM-None]
  b'hello'
# ---
# name: test_domain_unpack_command[packed0-SimpleCommandM]
  tuple(
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed0-SimpleCommandM].1
  tuple(
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed1-PingM]
  tuple(
    b'hello',
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed1-PingM].1
  tuple(
    b'hello',
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed10-ReadDowDeviceInformationM]
  tuple(
    1,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed10-ReadDowDeviceInformationM].1
  tuple(
    1,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed11-SetupTemperatureReportingM]
  tuple(
    list([
      1,
      2,
      3,
    ]),
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed11-SetupTemperatureReportingM].1
  tuple(
    list([
      1,
      2,
      3,
    ]),
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed12-DowTransactionM]
  tuple(
    1,
    3,
    list([
      0,
    ]),
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed12-DowTransactionM].1
  tuple(
    1,
    3,
    list([
      0,
    ]),
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed13-DowTransactionM]
  tuple(
    1,
    3,
    None,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed13-DowTransactionM].1
  tuple(
    1,
    3,
    None,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed14-SetupLiveTemperatureDisplayM]
  tuple(
    1,
    TemperatureDisplayItem(index=1, n_digits=3, column=1, row=1, units=<TemperatureUnit.FAHRENHEIT: 1>),
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed14-SetupLiveTemperatureDisplayM].1
  tuple(
    1,
    TemperatureDisplayItem(index=1, n_digits=3, column=1, row=1, units=<TemperatureUnit.FAHRENHEIT: 1>),
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed15-SendCommandToLcdControllerM]
  tuple(
    <LcdRegister.DATA: 0>,
    1,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed15-SendCommandToLcdControllerM].1
  tuple(
    <LcdRegister.DATA: 0>,
    1,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed16-ConfigureKeyReportingM]
  tuple(
    set({
      1,
    }),
    set({
      1,
    }),
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed16-ConfigureKeyReportingM].1
  tuple(
    set({
      1,
    }),
    set({
      1,
    }),
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed17-SetAtxPowerSwitchFunctionalityM]
  tuple(
    Functions enabled: KEYPAD_POWER_OFF
  Auto-Polarity Enabled: no
  Reset Inverted: yes
  Power Inverted: yes
  Power Pulse Length (seconds): 1.5,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed17-SetAtxPowerSwitchFunctionalityM].1
  tuple(
    Functions enabled: KEYPAD_POWER_OFF
  Auto-Polarity Enabled: no
  Reset Inverted: yes
  Power Inverted: yes
  Power Pulse Length (seconds): 1.5,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed18-ConfigureWatchdogM]
  tuple(
    2,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed18-ConfigureWatchdogM].1
  tuple(
    2,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed19-SendDataM]
  tuple(
    1,
    5,
    list([
      0,
    ]),
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed19-SendDataM].1
  tuple(
    1,
    5,
    list([
      0,
    ]),
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed2-WriteUserFlashAreaM]
  tuple(
    b'data',
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed2-WriteUserFlashAreaM].1
  tuple(
    b'data',
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed20-SetBaudRateM]
  tuple(
    115200,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed20-SetBaudRateM].1
  tuple(
    115200,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed21-SetGpioM]
  tuple(
    1,
    10,
    Function: 0
  Drive Mode: FAST_STRONG, RESISTIVE,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed21-SetGpioM].1
  tuple(
    1,
    10,
    Function: 0
  Drive Mode: FAST_STRONG, RESISTIVE,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed22-ReadGpioM]
  tuple(
    1,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed22-ReadGpioM].1
  tuple(
    1,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed3-SetLineM]
  tuple(
    b'hello',
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed3-SetLineM].1
  tuple(
    b'hello',
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed4-ReadLcdMemoryM]
  tuple(
    1,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed4-ReadLcdMemoryM].1
  tuple(
    1,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed5-SetCursorPositionM]
  tuple(
    1,
    1,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed5-SetCursorPositionM].1
  tuple(
    1,
    1,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed6-SetCursorStyleM]
  tuple(
    <CursorStyle.BLINKING_UNDERSCORE: 3>,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed6-SetCursorStyleM].1
  tuple(
    <CursorStyle.BLINKING_UNDERSCORE: 3>,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed7-SetContrastM]
  tuple(
    0.5,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed7-SetContrastM].1
  tuple(
    0.5,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed8-SetBacklightM]
  tuple(
    0.5,
    0.5,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed8-SetBacklightM].1
  tuple(
    0.5,
    0.5,
    None,
    None,
  )
# ---
# name: test_domain_unpack_command[packed9-SetBacklightM]
  tuple(
    0.5,
    None,
    0.25,
    3,
  )
# ---
# name: test_domain_unpack_command[packed9-SetBacklightM].1
  tuple(
    0.5,
    None,
    None,
    None,
  )
# ---
# name: test_domain_unpack_pack[-OptBytesM]
  None
# ---
# name: test_domain_unpack_pack[1-KeyPressM]
  1
# ---
# name: test_domain_unpack_pack[115200-BaudRateM]
  115200
# ---
# name: test_domain_unpack_pack[19200-BaudRateM]
  19200
# ---
# name: test_domain_unpack_pack[3-TemperatureDigitsM]
  3
# ---
# name: test_domain_unpack_pack[5-TemperatureDigitsM]
  5
# ---
# name: test_domain_unpack_pack[packed1-AtxPowerSwitchFunctionalitySettingsM]
  Functions enabled: KEYPAD_RESET
  Auto-Polarity Enabled: no
  Reset Inverted: yes
  Power Inverted: yes
  Power Pulse Length (seconds): 1.0
# ---
# name: test_special_character_pack_unpack
  tuple(
    'CFA533',
    'h1.4',
    'u1v2',
    b'\x00\x0e\x15\x15\x1f\x15\x1b\x0e',
  )
# ---
